frontend/model/captureLogs.js

```diff
@@ -0,0 +1,220 @@
+
+  localStorage.setItem('giConsole/lastEntry', logEntry)
+  localStorage.setItem('giConsole/count', entriesCount)
+
+  verifyEntryIsMarker()
+  verifyLogsSize()
+
+  // To avoid infinite loop because we log all selector calls, we run sbp calls
+  // here in a roundabout way by getting the function to which they're mapped.
+  // The reason this works is because the entire `sbp` domain is blacklisted
+  // from being logged in main.js.
+  sbp('sbp/selectors/fn', 'okTurtles.events/emit')(CAPTURED_LOGS, lastEntry)
+}
+
+function verifyEntryIsMarker () {
+  if (entriesCount % ENTRIES_MARKER_NTH === 0) {
+    const markers = getMarkers()
+    // Save entry as a marker to be later deleted when logs are too big.
+    markers.push(lastEntry)
+    localStorage.setItem('giConsole/markers', JSON.stringify(markers))
+  }
+}
+
+function verifyLogsSize () {
+  if (entriesCount >= ENTRIES_LIMIT) {
+    const markers = getMarkers()
+    let toDelete = ENTRIES_MARKER_NTH
+    let oldestEntry = markers.shift() // the oldest marker
+    do {
+      const log = JSON.parse(localStorage.getItem(`giConsole/${oldestEntry}`)) || {}
+      localStorage.removeItem(`giConsole/${oldestEntry}`)
+      entriesCount--
+      toDelete--
+      oldestEntry = log.prev
+    } while (toDelete && oldestEntry)
+
+    localStorage.setItem('giConsole/markers', JSON.stringify(markers))
+    localStorage.setItem('giConsole/count', entriesCount)
+
+    // Delete nth oldest logs recursively. This scenario can happen when the
+    // entries limit is changed. Example: There are 400 logs and the limit is 500.
+    // We change the limit to 100 and the marker nth to 25. 325 logs need to be deleted.
+    // We do it recursively in chunks of 25 until there's only 75 logs again.
+    verifyLogsSize()
+  }
+}
+
+function verifyLogsConfigs () {
+  // If ENTRIES_LIMIT or ENTRIES_MARKER_NTH are changed in a release
+  // we need to recalculate markers and verify if it reached the size limit.
+  const storedLimit = +localStorage.getItem('giConsole/limit')
+  const storedMarkerNth = +localStorage.getItem('giConsole/markerNth')
+
+  // when it's the first time running this, just set the correct limits
+  if (storedLimit === 0) {
+    localStorage.setItem('giConsole/limit', ENTRIES_LIMIT)
+    localStorage.setItem('giConsole/markerNth', ENTRIES_MARKER_NTH)
+    return
+  }
+
+  if (storedMarkerNth !== ENTRIES_MARKER_NTH) {
+    // recalculate markers based on new nth rule.
+    const newMarkers = []
+    let entryIndex = entriesCount
+    let prevEntry = lastEntry
+    do {
+      const log = JSON.parse(localStorage.getItem(`giConsole/${prevEntry}`))
+      if (!log) break
+      if (entryIndex % ENTRIES_MARKER_NTH === 0) {
+        newMarkers.push(prevEntry)
+      }
+      entryIndex--
+      prevEntry = log.prev
+    } while (prevEntry)
+    // reverse new markers to be in chronological order, and then store them.
+    localStorage.setItem('giConsole/markers', JSON.stringify(newMarkers.reverse()))
```

Comment:
I don't think this section is needed and I think it can be removed without any harm.

I think it's perfectly fine to replace lines 105 - 118 with simply:

```js
localStorage.setItem('giConsole/markers', JSON.stringify([]))
```

Then push to `newMarkers` not here, but inside of `captureLog()`.

Reply:
This section is responsible for redefining the markers when we change `ENTRIES_MARKER_NTH`. It's a needed operation, otherwise, we won't be able to identify the oldest entries and delete them later. This would lead us to have memory leaks on localStorage (too many entries and not being able to delete them properly)

Note: Greg, based on some of your latest reviews, I'm starting to believe you aren't running this code on your computer as I previously asked (and as it should always be done, especially in complex PR like this one). I highly recommend you do it before suggesting any major change like this one. If you do it, you'll know when a possible change breaks the code and avoid misleading and counterproductive comments.
Please, don't take this the wrong way. We are here together to build the best code possible but to do it so, I need you to do efficient code reviews.

Reply:
Thanks for the note of caution, you're right my head space has been too much in the coronavirus area (partly because of the quarantine and the hackathon). I did run the code before and that's how I noticed the issues before, but you're correct that I did not test these latest changes due to lack of time. I will do my best to get better about that! ^_^

You might be right about this... my thinking was that it might not be needed because the markers would simply be re-added, and sure it might go over the limit by a bit but not too much. However, you're right that depending on the numbers it could end up being too much, and this helps ensure consistent results. 👍

Reply:
Yeah... it's been a tough time for everyone :/

> the markers would simply be re-added

Yes, but it would be the newest logs, the oldest ones would be lost "forever" and we wouldn't be able to delete them later.

Reply:
> Yes, but it would be the newest logs, the oldest ones would be lost "forever" and we wouldn't be able to delete them later.

Why? It's a single chain. It doesn't matter where you put the marker. If you lost the previous markers, and add one later down the road, it's still pointing to the same section, the only difference is where the start/end of that section is.

Reply:
Yes it's a chain but it needs to be from the start. Let's imagine the following:

### Scenario
1000 logs limit. A marker every 250th log. There are 1050 logs. We would have 4 markers.
Markers: 1 (250th), 2 (500th), 3 (750th) and 4 (1000th).

### The limits are changed:
500 logs limit. A marker for each 50th log.

Now, 2 things can happen based on the code approach:

### A. If we reset the array (your suggestion):
The oldest logs won't be deleted because there are no markers, so we don't know what's the oldest log. So we'll keep having 1050 logs. When the 1100th log gets created, it will be turned into a marker. This will be the 1st marker. Then the condition to delete the oldest logs will be true and 50 logs will be deleted. But instead of deleting the oldest logs, it will delete the more recent logs because it's the only marker available. It's a bug!

### B. If we recalculate the markers (current code):
The oldest logs will be deleted because the markers were re-adjusted:
Markers: 1 (50th), 2 (100th), 3 (150th) ... 21(1050th).
After deleting all the oldest logs (with the while loop) there will be 450 logs and an array of 9 markers. When the 500th log happens, it will take the 1st marker and delete the oldest 50 logs as expected and everything is fine.

Explaining this is hard. It's way easier if you just test it by yourself on your computer.

Reply:
Yes, I think that sounds right. Your approach is more precise at the cost of being slightly more computationally expensive. That's all I was thinking. In Approach A, it would delete a whole bunch of logs, and reset everything back to normal. In Approach B, it would delete a lot of logs, but not as many. We can keep Approach B, it's totally fine. I'm just saying Approach A works too, it's just less precise.

Reply:
> I'm just saying Approach A works too.

It wouldn't work Greg. The oldest logs would be there forever. That's a major problem: reaching LS memory limit (9Mb). I can record a video for you to show you this.

Reply:
No worries, we'll do it your way!